lib/Optimizer/Transforms/LoopUnroll.cpp

```diff
@@ -157,18 +151,46 @@ class LoopUnrollPass : public cudaq::opt::impl::LoopUnrollBase<LoopUnrollPass> {
   using LoopUnrollBase::LoopUnrollBase;
 
   void runOnOperation() override {
-    auto *op = getOperation();
     auto *ctx = &getContext();
-    RewritePatternSet patterns(ctx);
-    bool passFailed = false;
-    patterns.insert<UnrollCountedLoop>(ctx, threshold, signalFailure,
-                                       passFailed);
-    if (failed(applyPatternsAndFoldGreedily(op, std::move(patterns))) ||
-        passFailed) {
+    auto *op = getOperation();
+    auto numLoops = countLoopOps(op);
+    if (numLoops) {
+      PassManager pm(ctx);
+      pm.addPass(createCanonicalizerPass());
+      RewritePatternSet patterns(ctx);
+      patterns.insert<UnrollCountedLoop>(ctx, threshold,
+                                         /*signalFailure=*/false);
+      FrozenRewritePatternSet frozen(std::move(patterns));
+      // Iterate over the loops until a fixed-point is reached. Some loops can
+      // only be unrolled if other loops are unrolled first and the constants
+      // iteratively propagated.
+      do {
+        (void)applyPatternsAndFoldGreedily(op, frozen);
+        if (failed(pm.run(op)))
+          break;
+      } while (loopsWereUnrolled(op, numLoops));
+    }
+    if (numLoops && signalFailure) {
```

Comment:
quick question - what is the reason for this last application of the unroll pass? Would this not have been unrolled fully by the above while loop?

Reply:
The condition means "if there are still loops in op and we're supposed to raise errors about them". So when that is true, we enable the error messages, run the pass one more time, and it spits out the errors.

We can't do it in the fixed-point processing, otherwise we get a slew of spurious reports on the intermediate cases that haven't yet been unrolled.